### PR TITLE
Update for <= k8s 1.12 kubernetes versions

### PR DESCRIPTION
**Pull request by LKaemmerling**
The cloud controller is now useable with <= k8s 1.12. We should release this as 1.3.0.

This was tested on K8s 1.12.

### CODE REVIEW

File: hcloud/instances.go [hunk 1]
```diff
@@ -88,15 +88,15 @@ func (i *instances) InstanceTypeByProviderID(providerID string) (string, error)
...
 	return errors.New("not implemented")
 }
 
-func (i *instances) CurrentNodeName(hostname string) (types.NodeName, error) {
+func (i *instances) CurrentNodeName(ctx context.Context, hostname string) (types.NodeName, error) {
 	return types.NodeName(hostname), nil
 }
 
-func (i instances) InstanceExistsByProviderID(providerID string) (exists bool, err error) {
+func (i instances) InstanceExistsByProviderID(ctx context.Context, providerID string) (exists bool, err error) {
```
Comment (thcyron) [inline hcloud/instances.go]: Now that you have a context, shouldn’t you pass it around when calling the hcloud client?

```go
server, _, err = i.client.Server.GetByID(context.Background(), id)
```

(A few lines below.)

File: hcloud/util.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func providerIDToServerID(providerID string) (id int, err error) {
 		return
 	}
 
-	idString := strings.Replace(providerID, providerPrefix, "", 1)
+	idString := strings.ReplaceAll(providerID, providerPrefix, "")
```
Comment (thcyron) [inline hcloud/util.go]: Just to make sure: these two lines don’t do the same (first line replaces first occurrence, second line replaces all occurrences).

File: hcloud/zones.go [hunk 3]
```diff
@@ -31,7 +33,8 @@ func newZones(client *hcloud.Client, nodeName string) cloudprovider.Zones {
 	return zones{client, nodeName}
 }
 
-func (z zones) GetZone() (zone cloudprovider.Zone, err error) {
+func (z zones) GetZone(ctx context.Context) (zone cloudprovider.Zone, err error) {
+	log.Print("GetZone")
```
Comment (thcyron) [inline hcloud/zones.go]: Do you want to keep those logs?

File: hcloud/util.go (hunk 2 above)
Comment (LKaemmerling) [reply hcloud/util.go]: This is intended.

File: hcloud/instances.go [hunk 4]
```diff
@@ -34,84 +34,100 @@ func newInstances(client *hcloud.Client) *instances {
...
 
-func (i *instances) NodeAddressesByProviderID(providerID string) ([]v1.NodeAddress, error) {
+func (i *instances) NodeAddressesByProviderID(ctx context.Context, providerID string) ([]v1.NodeAddress, error) {
 	id, err := providerIDToServerID(providerID)
 	if err != nil {
 		return nil, err
 	}
 
-	server, err := getServerByID(i.client, id)
+	server, err := getServerByID(i.client, id, ctx)
```
Comment (thcyron) [inline hcloud/instances.go]: FYI: It’s a generally-accepted convention to pass `ctx` always as the first argument.

**Pull request merged**